### ISSUE TITLE
Add Home Assistant support

### ISSUE THREAD

**Issue by popy2k14** [hyperion-project/hyperion.ng#1703]
## Feature request

It would be a nice addition if an native home assistant api could be implemented as an led hardware.

#### What problem does this feature solve?
This solves the issue of doing workarounds like emulate hue bridge in home assistant or something like this: https://community.home-assistant.io/t/use-home-assistant-lights-with-hyperion/678249/2

#### What does the proposed API look like?
See here, it's well documentated.: https://developers.home-assistant.io/docs/api/rest/

#### How should this be implemented in your opinion?
You could choose an entity for a single led on the led hardware setup.

#### Are you willing to work on this yourself?
No, sadly not.

**Comment by Lord-Grey** [comment]
What LED types are you addressing via HA?

I do not yet understand why an extra hop via HA is required or seem to be beneficial?
Running some LED updates directly and some others via HA will ensure that both updates are not on sync.
(Unless you run with very low fps).

**Comment by popy2k14** [comment]
I want address single ambient light bulbs, which are connected this way HA -> Zigbee2MQTT -> Zigbee network -> bulb.

The benefit of implementing HA is that you can address any HA supported "lights" which are a lot.
So with the middleware HA, hyperion is compatible with much much more lights.
Sure it's not realtime streaming, but on single ambient light bulbs (not strips with dozends of leds),
the update intervall must not be so high (in my case also a rgb wled strip (no daisy chain leds)) i have set "smoothing" in hyperion to 2000ms. So on single light bulbs update's could happen every 2000ms. The dimming/transition (in my case) is happaning on the bulbs itself.

**Comment by Lord-Grey** [comment]
Thank you that clarifies.

**Comment by popy2k14** [comment]
Thanks. Looking forward to this, so the whole HA community can benefit from this.

**Comment by kuba-orlik** [comment]
Another benefit of going through Home Assistant is that you can use the RGB lights both with hyperion, and with any other automations you might have set up in Home Assistant

**Comment by kuba-orlik** [comment]
As the author of the bridge I can say that most of the logic has been implemented [here](https://git.kuba-orlik.name/kuba/hyperion-home-assistant-bridge) - but I assume in order to be part of Hyperion, that would have to be implemented in c++, right?

**Comment by satgit62** [comment]
Hello,
I also find this project very useful and would be delighted if HA were implemented. It will expand the possibilities enormously.

**Comment by thueringer01** [comment]
Hello,
There are so many Homeassistant users worldwide. Many of them also use HyperHDR. I would also like to control my Zigbee lights with HyperHDR via Homeassistant. An integration of the kuba-orlik bridge into HyperHDR would be a major expansion for many users

**Comment by Paulchen-Panther** [comment]
@thueringer01 
Da bist du hier total falsch. Hyperhdr ist ein anderes Repository.

**Comment by satgit62** [comment]
> @thueringer01 Da bist du hier total falsch. Hyperhdr ist ein anderes Repository.

I think he uses HyperHDR, but it's important to him that Hyperion takes the first step. He would use what ZigBee lamps support, just like I do myself.

**Comment by thueringer01** [comment]
Sorry that I'm in the wrong place. Yes, I use Hyperhdr and Homeassistant. I would like a bridge from Hyperhdr to Homeassistant. With Homeassistant I can then control all the lights.

**Comment by popy2k14** [comment]
@Lord-Grey thanks a lot for implementing this! I think everyone with homeassistant is appreacciate this.
Sadly i just habe hyprion.ng running in webos on my LG: https://github.com/webosbrew/hyperion.ng-webos-loader

Maybe @throwaway96 can provide an LG webos build to test this?

**Comment by satgit62** [comment]
> I have not implemented features like a configuration wizard nor saving/restoring light state, as
> I would like to first check, if you find it usable in general.

Hello, 

I have tried to trigger actions to test Home Assistant Branch. Apparently something is wrong, because the actions are not executed successfully and are aborted with errors.
I then made a fork of their repositories and made the commits to be able to test the implementation of Home Assistant.  
I also tried to build “hype
[build_hyperion_ng.txt](https://github.com/user-attachments/files/16287375/build_hyperion_ng.txt)
[build_ipk.txt](https://github.com/user-attachments/files/16287376/build_ipk.txt)
rion.ng-webos-loader” for LG and took as reference the cloned repository with their commits. It was built including the new LED devices error free, however the daemon won't start on LG.
I am grateful for any hint. 
Here is the log of build_hyperion_ng and build_ipk.

**Comment by Lord-Grey** [comment]
@satgit62 It might be best you check at the WebOS Discord.
The core Hyperion team has no expertise on the WebOS.

PS: Let me merge the master into the HA branch, as there were some GitHub build related updates done.

**Comment by satgit62** [comment]
@Lord-Grey 
Okay, thanks very much, will try with the standard build.

**Comment by satgit62** [comment]
@Lord-Grey
Hello,
Thanks for the “**Merge branch ‘Lord-Grey:master’ into master**”.
Now everything worked. 
I even built “**org.webosbrew.hyperion.ng.loader**” for webOS, and it works as expected. The daemon started right away, I can select the registered lamps in Home Assistant, and they are also controlled by Hyperion. Wow!
This is a big upgrade for Hyperion. 
I will run more tests and get back to you. 
![webOS Hyperion NG Homeassistant_info](https://github.com/user-attachments/assets/e4449131-f7bc-4af8-bfc8-a855e1f04270)
![webOSHyperion NG Homeassistant_LED Controller](https://github.com/user-attachments/assets/98d63345-4085-4fa6-8ca2-1290b9826e84)

**Comment by popy2k14** [comment]
@satgit62 nice that you got it built for webos.
Any chance to get it shared for testing?

**Comment by satgit62** [comment]
> @satgit62 nice that you got it built for webos. Any chance to get it shared for testing?

Yes, of course, here is the link:
https://github.com/satgit62/hyperion.ng-webos-loader/actions/runs/9996016126/artifacts/1716243456

I would also appreciate your feedback.

**Comment by Lord-Grey** [comment]
> Now everything worked
@satgit62 That is great news!!!

**Comment by satgit62** [comment]
> > Now everything worked
> > @satgit62 That is great news!!!

You could really build it into the master branch of https://github.com/hyperion-project.
That will make a lot of home assistant users happy.
Like I said, I'll do some testing over the weekend and let you know if everything works as it should.

**Comment by Lord-Grey** [comment]
Happy to have it merged, after the weekend and further feedback.
In case some of the default values should change from a practical point of view, happy to have those updated too.
Have you tried to configure multiple HA Lights via one instance? Does that work too?

**Comment by satgit62** [comment]
> Happy to have it merged, after the weekend and get further feedback. In case some of the default values should change from a practical point of view, happy to have those updated too. Have you tried to configure multiple HA Lights via one instance? Does that work too?
Yes, I have also tested with two simultaneous Home Assistant ID and also an additional instance to my SK6812 RGBW CW LEDs Stripe and it works. I use, as main Ambilight instance, Adalight on HyperSerialPico with “Adafruit Feather RP2040 SCORPIO” LED controller. I will also run tests with WLED on ESP 32 and HyperSerial on ESP32 S2 Mini wemos and report back.

**Comment by satgit62** [comment]
> @satgit62 nice that you got it built for webos. Any chance to get it shared for testing?

@popy2k14
I think that was the wrong build, sorry. That was not yet merged by Lord Gray: “Merge branch ‘Lord-Grey:master’ into master”.

Here is the right link: https://github.com/satgit62/hyperhdr-webos-loader/releases/download/untagged-fc6b7947be214ad67bea/org.webosbrew.hyperion.ng.loader_0.1.11_all.zip

**Comment by Lord-Grey** [comment]
@satgit62 with this Hyperion code version you can also run HyperSerial, as well as ftdi devices.
Just give it a try….

**Comment by satgit62** [comment]
> @satgit62 with this Hyperion code version you can also run HyperSerial, as well as ftdi devices. Just give it a try….

Yes, I know, and FTDI LED-Device also works perfectly. I tested the one from @nurrik's build by heart back then.
But this one is up-to-date. I will test HyperSerial next.

**Comment by popy2k14** [comment]
@satgit62 thanks for the new build. Sadly i am on vacation and i need some time until i can test it.

**Comment by popy2k14** [comment]
@satgit62 found some time, sadly your last build link is not working.
Any chance to get a new one?

**Comment by popy2k14** [comment]
@satgit62 saw you did an dev pre release build here: https://github.com/satgit62/hyperion.ng-webos-loader/releases/
Is that the latest on which should be tested?

**Comment by popy2k14** [comment]
P.s.: can i just reinstall the new ipk with "webos-dev-manager"?
Because that's what i did?

**Comment by satgit62** [comment]
> P.s.: can i just reinstall the new ipk with "webos-dev-manager"? Because that's what i did?
> 
> P.S.P.S.: the ipk from here https://github.com/satgit62/hyperion.ng-webos-loader/releases/ does'nt work. hyperion deamon is just not starting. Reverted to the 2.0.16 from homebrew channel.
> 
> PLease let me know the test build, so i can test it.
> 
> thx

I realized it was the wrong link, but I wrote to them.
Try this one: https://github.com/satgit62/hyperion.ng-webos-loader/releases/download/untagged-2e2cbbc119a19994cf4b/org.webosbrew.hyperion.ng.loader_0.1.11_all.zip

**Comment by popy2k14** [comment]
> > P.s.: can i just reinstall the new ipk with "webos-dev-manager"? Because that's what i did?
> > P.S.P.S.: the ipk from here https://github.com/satgit62/hyperion.ng-webos-loader/releases/ does'nt work. hyperion deamon is just not starting. Reverted to the 2.0.16 from homebrew channel.
> > PLease let me know the test build, so i can test it.
> > thx
> 
> I realized it was the wrong link, but I wrote to them. Try this one: https://github.com/satgit62/hyperion.ng-webos-loader/releases/download/untagged-2e2cbbc119a19994cf4b/org.webosbrew.hyperion.ng.loader_0.1.11_all.zip

This link does'nt work for me, the message `Not Found` appears.
Maybe the build just works for your github account when you logged in?
Can you try the link in an private tab or other browser, when you not logged in?

thx

**Comment by satgit62** [comment]
> > > P.s.: can i just reinstall the new ipk with "webos-dev-manager"? Because that's what i did?
> > > P.S.P.S.: the ipk from here https://github.com/satgit62/hyperion.ng-webos-loader/releases/ does'nt work. hyperion deamon is just not starting. Reverted to the 2.0.16 from homebrew channel.
> > > PLease let me know the test build, so i can test it.
> > > thx
> > 
> > 
> > I realized it was the wrong link, but I wrote to them. Try this one: https://github.com/satgit62/hyperion.ng-webos-loader/releases/download/untagged-2e2cbbc119a19994cf4b/org.webosbrew.hyperion.ng.loader_0.1.11_all.zip
> 
> This link does'nt work for me, the message `Not Found` appears. Maybe the build just works for your github account when you logged in? Can you try the link in an private tab or other browser, when you not logged in?
> 
> thx

Have you logged in to GitHub?

**Comment by popy2k14** [comment]
Sure :-)
I meant, maybe it's an permission issue, so your private untagged build can't be access from other users?

**Comment by satgit62** [comment]
> Sure :-) I meant, maybe it's an permission issue, so your private untagged build can't be access from other users?

Yes, that's true. I'm on the road at the moment, and I'll fix it when I get back.

**Comment by popy2k14** [comment]
thx a lot. Just ping me.

**Comment by satgit62** [comment]
> thx a lot. Just ping me.

https://drive.google.com/file/d/1mNeRhlekHCW0BEI0fTaSrAeBTHeMa88X/view?usp=sharing

**Comment by thueringer01** [comment]
I tested the new Homeassistant integration today. Many thanks again to @Lord-Grey for the quick implementation. satgit62 took the build branch (Merge branch 'Lord-Grey:master' into master) as a reference, compiled it and built it for LG webOS in version 2.0.17-beta.1 . Runs very well. The RGB values ​​are transferred well to my Zigbee lamps. The brightness is always 255. The delay is acceptable. Can the value for the brightness be transferred? (in darkness = brightness 0, in fully white image = brightness 255?

**Comment by Lord-Grey** [comment]
@thueringer01 Could you share which supported color modes your devices have in HA?

Normally RGB has brightness included, therefore you might checking in Hyperion‘s Image Processing settings, if the config needs to be adapted.

I allow to a overwrite the brightness to set to a certain overall percentage, but that is applied ones an I would expect that it applied for all RGB updates later.

See also the [HA documentation](https://developers.home-assistant.io/docs/core/entity/light#color-modes) 

> Note that in color modes ColorMode.RGB, ColorMode.RGBW and ColorMode.RGBWW there is brightness information both in the light's brightness property and in the color. As an example, if the light's brightness is 128 and the light's color is (192, 64, 32), the overall brightness of the light is: 128/255 * max(192, 64, 32)/255 = 38%.

**Comment by thueringer01** [comment]
@Lord-Grey 
Unfortunately I'm only a tester and my English isn't that good, I use a translator.
The hyperhdr log says, for example:
`2024-07-24T19:04:49.398Z [LEDDEVICE|Homeassistant Lampe rechts] (DEBUG) (ProviderRestApi.cpp:274:executeOperation()) POST took 56ms, HTTP 200: [http://192.168.220.15:8123/api/services/light/turn_on] [{"entity_id":["light.0x04cd15fffea26a11","light.0x04cd15fffe8de6e9"],"rgb_color":[50,14,34]}]`
My zigbee lamp has the following attributes in Homeassistant:
in_color_temp_kelvin: 2000
max_color_temp_kelvin: 6535
min_mireds: 153
max_mireds: 500
effect_list: blink, breathe, okay, channel_change, finish_effect, stop_effect, colorloop, stop_colorloop
supported_color_modes: color_temp, xy
effect: null
color_mode: xy
**brightness: 255**
color_temp_kelvin: null
color_temp: null
hs_color: 15.259, 90.98
rgb_color: 255, 82, 23
xy_color: 0.66, 0.328
Brightness is always at 255.The other attributes are not important.

**Comment by Lord-Grey** [comment]
@thueringer01 Ich kann Dir dazu leider nichts sagen, da ich den HyperHDR Code nicht kenne.
Ich weiß nicht, wie der Hyperion Code dort angepasst wurde.

**Comment by satgit62** [comment]
@Lord-Grey
Es gibt keine Implementierung/Anpassung des Home Assistant in HyperHDR.
@Thueri01 meint sicherlich, **Hyperion log** und nicht **HyperHDR**.
Er hat bis vor kurzem HyperHDR benutzt. Da er Home Assistant verwendet und dort mehrere Leuchten betreibt, testet er derzeit Hyperion.

**Comment by Lord-Grey** [comment]
@thueringer01 Nichts für ungut. Danke fürs Testen. Ich muss mir mal anschauen, wie sich XY Lights verhalten.
Welches HA Light Plugin für Deine Lights benutzt Du?

**Comment by thueringer01** [comment]
@Lord-Grey 
Ja, das ist ein Hyperion Log, benutze sonst HyperHDR. Sorry für ungenaue Info.
In HA nutze ich den Mosquitto broker und Zigbee2MQTT. Als Bridge nutze ich einen Silicon_Labs_CP2102N_USB_to_UART_Bridge_Controller. 
Wenn möglich helfe ich gern mit. Kann leider nur testen.

**Comment by satgit62** [comment]
> thx a lot. Just ping me.

@popy2k14
Hello,
were you able to test my webOS build successfully?

**Comment by popy2k14** [comment]
@satgit62 Sorry for the delay.
Found time to test your latest ipk (from gdrive).
Sadly the hyperion.ng service does'nt start on my OLED65B19LA, here is the syslog:

```
[15:21:45.957][INFO] jailer JAILER_INFO file unlock appid: org.webosbrew.hyperion.ng.loader {}
[15:21:45.957][WARNING] jailer JAILER_INFO app launch exec(/media/developer/apps/usr/palm/services/org.webosbrew.hyperion.ng.loader.service/loader_service,  /media/developer/apps/usr/palm/services/org.webosbrew.hyperion.ng.loader.service/loader_service) {}
[15:21:45.976][INFO] hyperion.ng-loader INFO main: Starting up... {}
[15:21:45.985][WARNING] <default-lib> LS_REQUIRES_SECURITY Service security groups don't allow method call. {"SERVICE":"org.webosbrew.hyperion.ng.loader.service","CATEGORY":"/power","METHOD":"getPowerState"}
[15:21:45.985][ERR] <default-lib> LS_NO_METH Couldn't find method: getPowerState {"METHOD":"getPowerState"}
[15:21:45.985][INFO] hyperion.ng-loader INFO power_callback: Power status callback message: {"returnValue":false,"errorCode":-1,"errorText":"Denied method call \"getPowerState\" for category \"/power\""} {}
[15:21:46.507][WARNING] <default-lib> LS_DEPRECATED Deprecated method call {"SENDER":"com.webos.app.screensaver-5013","SERVICE":"com.webos.app.screensaver-5013","CATEGORY":"/time","METHOD":"getEffectiveBroadcastTime","FLAGS":1}
[15:21:47.541][WARNING] <default-lib> LS_DEPRECATED Deprecated method call {"SENDER":"com.webos.app.screensaver-5013","SERVICE":"com.webos.app.screensaver-5013","CATEGORY":"/time","METHOD":"getEffectiveBroadcastTime","FLAGS":1}
[15:21:48.068][INFO] hyperion.ng-loader INFO signal_handler: SIGNAL=15 detected, stopping... {}
[15:21:48.575][WARNING] <default-lib> LS_DEPRECATED Deprecated method call {"SENDER":"com.webos.app.screensaver-5013","SERVICE":"com.webos.app.screensaver-5013","CATEGORY":"/time","METHOD":"getEffectiveBroadcastTime","FLAGS":1}
[15:21:49.640][WARNING] <default-lib> LS_DEPRECATED Deprecated method call {"SENDER":"com.webos.app.screensaver-5013","SERVICE":"com.webos.app.screensaver-5013","CATEGORY":"/time","METHOD":"getEffectiveBroadcastTime","FLAGS":1}
[15:21:49.823][INFO] jailer JAILER_INFO file lock appid: org.webosbrew.hyperion.ng.loader {}
```

the above log is looping, the tv is trying to start hyperion.ng over and over again.

When i go back to the current 2.0.16 from homebrew channel it's working again.

Any hint's

**Comment by popy2k14** [comment]
I think this is the issue:

```
[15:21:45.985][WARNING] <default-lib> LS_REQUIRES_SECURITY Service security groups don't allow method call. {"SERVICE":"org.webosbrew.hyperion.ng.loader.service","CATEGORY":"/power","METHOD":"getPowerState"}
[15:21:45.985][ERR] <default-lib> LS_NO_METH Couldn't find method: getPowerState {"METHOD":"getPowerState"}
[15:21:45.985][INFO] hyperion.ng-loader INFO power_callback: Power status callback message: {"returnValue":false,"errorCode":-1,"errorText":"Denied method call \"getPowerState\" for category \"/power\""} {}
```

My guess is that there is some define you have to check/correct regarding to "getPowerState"?
Maybe @throwaway96 can help here?

Would be nice to have an working build to test before merging.
thx

**Comment by satgit62** [comment]
> I think this is the issue:
> 
> ```
> [15:21:45.985][WARNING] <default-lib> LS_REQUIRES_SECURITY Service security groups don't allow method call. {"SERVICE":"org.webosbrew.hyperion.ng.loader.service","CATEGORY":"/power","METHOD":"getPowerState"}
> [15:21:45.985][ERR] <default-lib> LS_NO_METH Couldn't find method: getPowerState {"METHOD":"getPowerState"}
> [15:21:45.985][INFO] hyperion.ng-loader INFO power_callback: Power status callback message: {"returnValue":false,"errorCode":-1,"errorText":"Denied method call \"getPowerState\" for category \"/power\""} {}
> ```
> 
> My guess is that there is some define you have to check/correct regarding to "getPowerState"? Maybe @throwaway96 can help here?
> 
> Would be nice to have an working build to test before merging. thx

@popy2k14, It works as expected for other users who have also tested it.
They have certainly not done a proper reboot. Quick Start+ must be switched off or simply disconnect the device from the power supply.

**Comment by popy2k14** [comment]
@satgit62 on my tv the deamon does'nt start.
QuickStart+ is disabled, also rebootet the tv several times (als cold restart with power plug).

![PXL_20240727_142744125](https://github.com/user-attachments/assets/f50e81bb-70e4-4383-aa77-02222a387c63)
![PXL_20240727_143106891](https://github.com/user-attachments/assets/62035106-1884-4a5f-aac0-286c4f414262)
![PXL_20240727_143140470](https://github.com/user-attachments/assets/3ef8fe61-f24b-40bd-a141-268c56e946be)

I am uninstalling/installing the ipk with WebOS dev manager.
After uninstalling the app is also gone from the main menu, and returns after using your ipk.
So i think my method of un-/installing is working.
But the daemon is crashing/stopping regardsing the missing function call!?

Any hint's?

**Comment by popy2k14** [comment]
ok, i guess it has nothing todo with the getPowerState mentioned before.
There is no return "false" in the code:

![image](https://github.com/user-attachments/assets/b4923dec-585f-4088-97c0-9753a5352c7d)

Any hint's?
How do you install the ipk? ALso with "WebOS dev manager"?

**Comment by satgit62** [comment]
> ok, i guess it has nothing todo with the getPowerState mentioned before. There is no return "false" in the code:
> 
> ![image](https://private-user-images.githubusercontent.com/6324597/352760263-b4923dec-585f-4088-97c0-9753a5352c7d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjIwOTMzMzAsIm5iZiI6MTcyMjA5MzAzMCwicGF0aCI6Ii82MzI0NTk3LzM1Mjc2MDI2My1iNDkyM2RlYy01ODVmLTQwODgtOTdjMC05NzUzYTUzNTJjN2QucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDcyNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA3MjdUMTUxMDMwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9Nzg2NWQzN2MxYWZkNzM1MzM4OThjMDgxMDkwMDNkYzQyYjAwNTczMmFmMzRjNDJlMzQwNDk2ODQ0N2JhZTI0MiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.ri_aGYM4AZriw0S6QAmSKVChB8kVTr6fmsjWxeMbYAk)
> 
> Any hint's? How do you install the ipk? ALso with "WebOS dev manager"?

I uninstalled as normal via webOS Dev Manager, rebooted, installed the new ipk via webOS Dev Manager, and called it from there and waited until the services had “elevated”. Then switched on Autostart, pressed Start and rebooted. If there had been an error in the code, then it would not have worked for me and the other users either.

Can you compare whether the content is the same for you? 
![org webosbrew hyperion ng loader services-hyperion](https://github.com/user-attachments/assets/a0869b58-dc11-40b1-861a-92da81e2ee7b)

**Comment by popy2k14** [comment]
@satgit62 thx a lot, it's working now.
I missed the reboot after uninstalling the original version.
Now the daemon is started and i configure and test it.
Will come back here.

**Comment by satgit62** [comment]
@popy2k14
I recommend that you save the settings after each Home Assist instance and restart before creating a new instance and so on. The origin branch of @Lord-Grey has not been finalized or merged into the final Hyperion Master.

**Comment by popy2k14** [comment]
@Lord-Grey @satgit62 i have now tested the home assistant build, and this are my results.

Sometimes hyperion-ng crashes back to the login page (when fiddling around with the home assistant instance).
two times i could not log in back -> "Authorization error". I had to stop and start the instance to get it working again.

I have tested three different home assistant light types:

1. HA -> MQTT -> Z2M -> Zigbee -> hue light strip
    It worked right out of the box but color changes where choppy (which is distracting during watching).
    There is an "transition" flag which could be given on color changes, so it will fade on it's own.
    Maybe that should be an option in the implementaion (or locktime (Sperrzeit in german) could be used)

2. HA -> ESPhome -> 5 channel light strip (RGBWW)
    It worked right out of the box and also fading between color is perfect :-)

3. HA -> ESPhome -> 59x WS2812 Led's (RGB)
    If the lock time (Sperrzeit) is the default of 50ms, the leds does nothing.
    Hd to raise it to ~700ms for it to react, but also than it's slow.
    I know that's expected, because esphome firmware is busy controlling all the 59x led's.
    So such devices should not be used in this way :-)

So all in all i am impressed with the first implementation.
Maybe 1. could be improved.

**Comment by satgit62** [comment]
> Sometimes hyperion-ng crashes back to the login page (when fiddling around with the home assistant instance). two times i could not log in back -> "Authorization error". I had to stop and start the instance to get it working again.

@popy2k14, as I said, it's a first attempt that looks very promising. @Lord-Grey also mentioned that this is a '**first implementation of creating a home assistant light feed**'. Also in https://github.com/hyperion-project/hyperion.ng/pull/1763, he said, “!!! The implementation is a proof of concept
A list of HA lights can currently be mapped to the first layout area.
For further coverage areas, an additional instance would be required.
It is recommended to set the latency so that a certain amount of time passes between two updates.
The solution cannot run at high fps and should only be used for individual lights, i.e., not for the main lights.”

Development like this takes time to get right, and it doesn't happen overnight. I'm glad @Lord-Grey is supporting this project at all.

**Comment by popy2k14** [comment]
Don't get me wrong. 
Nice that the first step is done, thx a lot @Lord-Grey !!!!
I just reported what I discovered 😊

**Comment by satgit62** [comment]
@popy2k14

Hello,

I took my time today and built a new “hyperion.ng-webos-loader”. I have slimmed down this version by 20 MB and removed various bugs.
Furthermore, I would be happy if you could test it. It is best to delete the old Home Assistant instances and create a new one. Hostname/IP-address=homeassistant.local or the actual IP address and port=8123. Everything else as before. Don't forget the restarts.
Please give me feedback on whether this has worked.

https://github.com/satgit62/hyperion.ng-webos-loader/releases/download/latest/org.webosbrew.hyperion.ng.loader_0.1.11_all.ipk  
![homeassistant local](https://github.com/user-attachments/assets/7a768294-0b99-4cf6-ad94-ba76d42c5e6f)

**Comment by popy2k14** [comment]
@satgit62 sorry for the delay, sadly the build is not "complettly" working.
My previously configured led instance is working (see picture),
but the WebUI is not working (see pictures).

![dashboard not working](https://github.com/user-attachments/assets/6de43bb0-f99b-46a9-bd18-814fae1021c9)
![not_running](https://github.com/user-attachments/assets/08a9d29e-1d9f-4977-b813-38a73c760ec1)
![PXL_20240811_094620654](https://github.com/user-attachments/assets/34a6d67c-41a7-482f-a823-473126565b99)

**Comment by popy2k14** [comment]
@satgit62 goit it working!
It seems some cookies leads to this issue, see here: https://github.com/hyperion-project/hyperion.ng/issues/1357
New build is runing now (with WebUI) and i will test it and come back here.

**Comment by popy2k14** [comment]
@satgit62 Wanted to setup HA (like i did in your previous build) but have issues!

on first setup, i have NO "port" fieldsee screenshot:
![port_setting_missing](https://github.com/user-attachments/assets/d594b99f-a3e8-4329-b5c7-43ea5582915a)

After Saving the instance without "port", WebGUI shows empty configuration and also "homeassisant" device is gone, see screenshot:
![HA_missing](https://github.com/user-attachments/assets/c302b89a-eb93-4b8b-8724-0225b2b984cd)

Also in the logs there is this message:

`2024-08-11T10:34:32.959Z [LEDDEVICE|HomeAssistant] (ERROR) Invalid port [0] for host: (192.168.0.8)! - Port must be in range [1 - 65535]`

Same issue with your test build on gdrive.
I have than exported the config, corrected the port, which was:

`		"port": 0,`

to

`		"port": 8123,`

and now its connected and i will do further tests.
But port will not be shown in the GUI.

Is this intentional?

**Comment by satgit62** [comment]
> @satgit62 Wanted to setup HA (like i did in your previous build) but have issues!
> 
> on first setup, i have NO "port" fieldsee screenshot: ![port_setting_missing](https://private-user-images.githubusercontent.com/6324597/356867704-d594b99f-a3e8-4329-b5c7-43ea5582915a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjMzNzUzODYsIm5iZiI6MTcyMzM3NTA4NiwicGF0aCI6Ii82MzI0NTk3LzM1Njg2NzcwNC1kNTk0Yjk5Zi1hM2U4LTQzMjktYjVjNy00M2VhNTU4MjkxNWEucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDgxMSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA4MTFUMTExODA2WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZmMzMTIzZmFhNzNhMzVhNjI0YjYwMGU3NDZmODUxMGMyOWVhOWUxMTJiMmExOThkNmEyMDZmMTNjOGQwYTljZSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.5jL0oZPT8XMxjr-bWtalt3BdJpgHUDJQcu-Fp8rjyAs)
> 
> After Saving the instance without "port", WebGUI shows empty configuration and also "homeassisant" device is gone, see screenshot: ![HA_missing](https://private-user-images.githubusercontent.com/6324597/356867739-c302b89a-eb93-4b8b-8724-0225b2b984cd.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjMzNzUzODYsIm5iZiI6MTcyMzM3NTA4NiwicGF0aCI6Ii82MzI0NTk3LzM1Njg2NzczOS1jMzAyYjg5YS1lYjkzLTRiOGItODcyNC0wMjI1YjJiOTg0Y2QucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDgxMSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA4MTFUMTExODA2WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ODdkNmE5ODY2ZjA1Y2JkMjg3YjhmZTA4ZjQwZGRmYmM5ZmI1MDg0ZTg1OWYxZWJlZmE3OTI4MDM2ZjdkMjE3ZCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.WxLOj0iPO-4_woeNBsXD7D9s6YRpSQe76SlpyAoozjA)
> 
> Also in the logs there is this message:
> 
> `2024-08-11T10:34:32.959Z [LEDDEVICE|HomeAssistant] (ERROR) Invalid port [0] for host: (192.168.0.8)! - Port must be in range [1 - 65535]`
> 
> Same issue with your test build on gdrive. I have than exported the config, corrected the port, which was:
> 
> ` "port": 0,`
> 
> to
> 
> ` "port": 8123,`
> 
> and now its connected and i will do further tests. But port will not be shown in the GUI.
> 
> Is this intentional?

You need to reboot your TV properly or disconnect it from the mains.
You can also enter homeassistant.local instead of IP address. See picture. I have already explained how to install this in: https://github.com/hyperion-project/hyperion.ng/issues/1703#issuecomment-2254175124
![homeassistant local](https://github.com/user-attachments/assets/9506c4af-a5e3-4ce7-a17e-6bf67115a2e4)

**Comment by popy2k14** [comment]
I have rebooted several times without luck.
Think it's an issue with the `~/.hyperion/db/hyperion.db` settings file.
Will delete it later and reimport the config and hope it fixes it.

Playing now with the settings and some HA lights.

**Comment by satgit62** [comment]
> I have rebooted several times without luck. Think it's an issue with the `~/.hyperion/db/hyperion.db` settings file. Will delete it later and reimport the config and hope it fixes it.
> 
> Playing now with the settings and some HA lights.

You must not import the old HA instances.
It is best to delete the old Home Assistant instances and create a new one. Hostname=homeassistant.local or the current IP address and port=8123. Everything else as before. Do not forget the restarts.

**Comment by popy2k14** [comment]
> > I have rebooted several times without luck. Think it's an issue with the `~/.hyperion/db/hyperion.db` settings file. Will delete it later and reimport the config and hope it fixes it.
> > Playing now with the settings and some HA lights.
> 
> You must not import the old HA instances. It is best to delete the old Home Assistant instances and create a new one. Hostname=homeassistant.local or the current IP address and port=8123. Everything else as before. Do not forget the restarts.

I have done this already and removed it from the exported json, reimported but it was still here.
Hyperion.ng on webos is storing it's settings in `~/.hyperion/db/hyperion.db`
Will later remove this database and reimport an config without homeassistant instance and recreate it.

**Comment by popy2k14** [comment]
@satgit62 @Lord-Grey There is a bug in the new implementation for homeassistant.
I have deleted the settings db (so hyperion is like fresh installed) and also than there are settings missing in the web gui.
Ps.: i have NOT configured anything else after factory reset or restored something!

Here is a screenshot how it looks like on my TV's installation (after deleteting hyperion db):
![image](https://github.com/user-attachments/assets/47c6bad8-10a9-4208-9904-02ee9e9ae83c)

And this is a modified screenshot of @satgit62 how it should be look like:
![should_look_like](https://github.com/user-attachments/assets/5e0ad361-5fe1-46f4-bf77-71d4b4b2c73b)

The version clearly missing:

- Port
- Override Brightness
- Brightness

Maybe because my browser is set to "german" language?
Can you please look into it?

**Comment by Paulchen-Panther** [comment]
@popy2k14 
Have you set the settings level to Advanced?

**Comment by popy2k14** [comment]
@Paulchen-Panther man i am blind :-) yep, that was it. thx a lot

@satgit62 @Lord-Grey the default port for HA was set to 0 but should be 8123. When someone (like me :-) has default settings level, than the homeassistant integration could not work beause of that.

I have now setup 3x lamps in HA and see how it goes.

It would be nice to have a override brightness for every ligth entity.
Now i have three instances dedicated to HA, just that i can set individual override brightness for every light.

**Comment by satgit62** [comment]
> @Paulchen-Panther man i am blind :-) yep, that was it. thx a lot

This is not a reproach, but you have to go into expert mode before you start with the settings.
You try to put everything into practice too quickly and forget elementary things in the process. I'm beginning to wonder why it works for others and not for you. I can set the brightness for each instance individually.
But joking aside, it's nice that it worked for you too.

**Comment by popy2k14** [comment]
> > @Paulchen-Panther man i am blind :-) yep, that was it. thx a lot
> 
> This is not a reproach, but you have to go into expert mode before you start with the settings. You try to put everything into practice too quickly and forget elementary things in the process. I'm beginning to wonder why it works for others and not for you. I can set the brightness for each instance individually. But joking aside, it's nice that it worked for you too.

Yeah, sometims you search an look and it's so obvious, anyway :-)

But the two things i mentioned should be addressed:

- port was 0 (and default not visible), in my case
- yes, i can set the brightness for each instance but it would be nice to set it for every entity/id of each instance

thx

**Comment by satgit62** [comment]
* yes, i can set the brightness for each instance but it would be nice to set it for every entity/id of each instance

@popy2k14 
This is exactly what works. I have three Homeassistant instances set up, and when I select “Override brightness”, I can set the brightness for each instance.
![Homeassistant Instace Smoking](https://github.com/user-attachments/assets/5ad9a02e-0bf7-4027-b028-72eb258e6aa6)
![Homeassistant Instace SofaWZ](https://github.com/user-attachments/assets/b3df4a6e-b13f-41f1-943c-2a2dbc0fa2fd)
![homeassistant Instance Bloom](https://github.com/user-attachments/assets/c7c349ee-b11b-44f0-8469-2641ac52c9e8)

**Comment by popy2k14** [comment]
@satgit62 i know :-) but it would be nice if a user adds two or more lights (entitys are they called in HA) to one instance, it would be nice if the user can set the override for every entity. In my case i want to add two lights with from different manufacturers and set different brightness.

But sure, i did it like you and have now three more instances, each set with different brightness.

**Comment by satgit62** [comment]
> @satgit62 i know :-) but it would be nice if a user adds two or more lights (entitys are they called in HA) to one instance, it would be nice if the user can set the override for every entity. In my case i want to add two lights with from different manufacturers and set different brightness.
> 
> But sure, i did it like you and have now three more instances, each set with different brightness.

@popy2k14 
This is not yet possible. Please also note that the HA Light IDs can currently only be assigned to the first LEDs in the layout. See: https://github.com/hyperion-project/hyperion.ng/pull/1763#issuecomment-2241735189
No idea if this is possible. Wait and see what @Lord-Gray has to say.

**Comment by Lord-Grey** [comment]
> the default port for HA was set to 0 but should be 8123. When someone (like me :-) has default settings level, than the homeassistant integration could not work because of that.

@popy2k14 That is my bad! I missed to set the default port in the UI schema.
I only tested until now using mDNS (those names with .local at the end) and there the port is resolved automatically.
I fixed that one already.

> there are settings missing in the web gui.

@popy2k14 You already found out...
I try to only show the minimum setting required in default mode to ensure that a standard user can use a devise best with just saving it. Here of course a token is required.
Unfortunately, this does not work, if I miss to provide the right defaults 8see above) :)

> It is best to delete the old Home Assistant instances and create a new one. ...
> You need to reboot your TV properly or disconnect it from the mains.

@satgit62 Not sure why to reboot. a configuration change should work without reboot.
In addition, if you would like to start from scratch for a given instance configuring the device,
just change the led device to type "file" and save.
If you then select homeassistant again, the configuration starts from square one.
That is the easiest way to restart configurationand get the (new) defaults.

@satgit62, @popy2k14 
In addition, I change the default latchtime (250ms) and the allowed range (100-2000ms).
In case the default time between two calls is still too low, happy to get feedback how long calls take in your environment.
For me one WLED call via HA takes ~150-300ms (sometimes more

**Comment by Lord-Grey** [comment]
> but it would be nice if a user adds two or more lights (entitys are they called in HA) to one instance, it would be nice if the user can set the override for every entity. In my case i want to add two lights with from different manufacturers and set different brightness.

@popy2k14 I understand where you are coming from...

Before going into possible ways of solutioning an HA light feed, you need to understand what is possible via the HA API.

Via one API call you can update 1 to n entities with the same color and/or brightness, [see here](https://www.home-assistant.io/integrations/light). One API call takes a good amount of time (~250ms or even more).

Given the above there are two options to run HA devices using the HA API.

1. You configure all lights via one instance and every light has its own brightness.

Pros: 
- It is simple from a User configuration perspective

Cons: 
- For every light configured an individual call is to be done in sequence.
i.e.to complete an update sequence n times 250ms is required.
- If an API call fails in between the whole device would fail.
- Currently the brightness is anyway only set when turning on the device to ensure that a previously dimmed device could use the full range during updates. All devices will get color updates according to the brightness definitions under "Image Processing"

2. You configure lights that would pick up the same layout area (and are of same type) via an individual instance (current implementation)

Pros:
- Only one call by Hyperion, HA handles the devices of the request
- You the device instance specific process can be configure via Image processing focussing on the device.
-  API requests for all instances are executed in parallel and not in sequence.

Cons:
- There is more one time configuration effort.

Hoping the above provide more insight into the topic.
Happy to get feedback, if the focus is more on the usability (1) or processing aspect (2).

@popy2k14 After writing the above, I am still not clear what is the expectation on brightness per light.
I assume you interpret brightness as the brightness during process, while it is the brightness the lights in an instance are powered on with, i.e. it defines the maximum brightness that can be achieved.

**Comment by satgit62** [comment]
@Lord-Grey

Thank you for the detailed explanation and further development of this project. You are right, Hyperion does not need to be restarted to configure and activate a new instance. Everything works fine on my Raspberry Pi and also under Windows. However, Hyperion.NG sometimes behaves strangely on webOS and does not react as expected. By restarting the TVs, the root services are elevated again under webOS and the Hyperion daemon is started cleanly. 
I will continue to test the two options you mentioned and report back to you. Many thanks again.

**Comment by Lord-Grey** [comment]
> I will continue to test the two options you mentioned

@satgit62 Please note that I have only implemented option 2 currently!
You can define multiple entries, but they are mapped to one layout area and updated via one call to HA.

I have done some updates to address the port issue etc., but have not pushed them yet. I am doing some additional enhancements and push all changes in a couple of days.
The you would not need to do a piece by piece testing on the updates.

**Comment by popy2k14** [comment]
> > the default port for HA was set to 0 but should be 8123. When someone (like me :-) has default settings level, than the homeassistant integration could not work because of that.
> 
> @popy2k14 That is my bad! I missed to set the default port in the UI schema. I only tested until now using mDNS (those names with .local at the end) and there the port is resolved automatically. I fixed that one already.

Thanks a lot for fixing this.

> 
> > there are settings missing in the web gui.
> 
> @popy2k14 You already found out... I try to only show the minimum setting required in default mode to ensure that a standard user can use a devise best with just saving it. Here of course a token is required. Unfortunately, this does not work, if I miss to provide the right defaults (see above) :)

Yeah, i was a user searching for the port ;-)
But i understand the settings level.
Would be nice if they where saved per instance and not in the browser cookies.

> @satgit62, @popy2k14 In addition, I change the default latchtime (250ms) and the allowed range (100-2000ms). In case the default time between two calls is still too low, happy to get feedback how long calls take in your environment. For me one WLED call via HA takes ~150-300ms (sometimes more

Tha'ts a good change. I found out that (without reading here) have also set 250ms which worked most of the time.
50ms was way to fast.

**Comment by popy2k14** [comment]
> > but it would be nice if a user adds two or more lights (entitys are they called in HA) to one instance, it would be nice if the user can set the override for every entity. In my case i want to add two lights with from different manufacturers and set different brightness.
> 
> @popy2k14 I understand where you are coming from...
> 
> Before going into possible ways of solutioning an HA light feed, you need to understand what is possible via the HA API.
> 
> Via one API call you can update 1 to n entities with the same color and/or brightness, [see here](https://www.home-assistant.io/integrations/light). One API call takes a good amount of time (~250ms or even more).
> 
> Given the above there are two options to run HA devices using the HA API.
> 
> 1. You configure all lights via one instance and every light has its own brightness.
> 
> Pros:
> 
> * It is simple from a User configuration perspective
> 
> Cons:
> 
> * For every light configured an individual call is to be done in sequence.
>   i.e.to complete an update sequence n times 250ms is required.
> * If an API call fails in between the whole device would fail.
> * Currently the brightness is anyway only set when turning on the device to ensure that a previously dimmed device could use the full range during updates. All devices will get color updates according to the brightness definitions under "Image Processing"
> 
> 2. You configure lights that would pick up the same layout area (and are of same type) via an individual instance (current implementation)
> 
> Pros:
> 
> * Only one call by Hyperion, HA handles the devices of the request
> * You the device instance specific process can be configure via Image processing focussing on the device.
> * API requests for all instances are executed in parallel and not in sequence.
> 
> Cons:
> 
> * There is more one time configuration effort.
> 
> Hoping the above provide more insight into the topic. Happy to get feedback, if the focus is more on the usability (1) or processing aspect (2).
> 
> @popy2k14 After writing the above, I am still not clear what is the expectation on brightness per light. I assume you interpret brightness as the brightness during process, while it is the brightness the lights in an instance are powered on with, i.e. it defines the maximum brightness that can be achieved.

Thanks for clearing things up regarding the HA API.
When i got you right:
So it's impossible currently to send a max brightness value to each entity, this would double the API calls,
so the user can just create another instance.

In my case i have lights from different manufactures which are positioned similar in the room.
One is brighter as the other with the same data, that's the reason i was asking for that.

I was also asking for that because i wanted to have as less instances in Hyperion as possible.
I dont know the inner workings of Hyperion, but from an performance perspective, having more instances means more cpu load 
(because of the pixel calculations and filters, ...).
Please correct me if i am wrong.

PS.: Sadly i have to disable the other ambilights (throught HA) because my wife found it distracting.
And you know: Happy wife, happy life :-)

**Comment by Lord-Grey** [comment]
Hi all

I have now pushed an updated version.

It includes the default port fix, as well as an updated default latch-time.
In addition, you can select your HA lights set-up at your HA server.
(in case there are non working lights, please come back to me, that we can filter them out from the list).
With the wizard button you can define the HA lights' position in the layout.

For testing, I would suggest you change your current HA config to a "File" device (do not for get to copy the token before).
Then you can select the "homeassistant" controller and start using the new features and updated default values.

A step by step guidance you find in the #1763 PR description.

**Comment by Lord-Grey** [comment]
> There is an "transition" flag which could be given on color changes, so it will fade on it's own.

@popy2k14 I wanted to make you transition time available, but unfortunately the time can only be given in a multiple of one second, which I do not think makes sense. I have seen that for some devices, e.g. Yeelight one could define the time in ms in HA.
For Hue I did not find that parameter...

**Transition**

>  	Number that represents the time (in seconds) the light should take to transition to the new state.

see https://www.home-assistant.io/integrations/light

**Comment by Lord-Grey** [comment]
Hi
further updates done...

On brightness, 3 configuration elements are now provided:

1&2: "Overwrite Brightness" & "Brightness value"
On: Sets "Brightness value" for every update, i.e. even if a user changes brightness in HA, the next update will set brightness to the configured value again.
OFF: A brightness change by user via HA will take place and brightness stays at the given value.

3. "Full Brightness at start" will set the brightness to 100% when HA streaming is powered on.
If a brightness change by user via HA afterwards will take place then brightness stays at the given value.
This ensures that at start the full colour range will be used

**Comment by satgit62** [comment]
Hi @all,
here is the link to the LG webOS pre-test version:https://drive.google.com/file/d/1t9fRLavJAIzMPtsEb1h7Ox_C5IUjmnqQ/view?usp=sharing

**Comment by IamDan77** [comment]
> Hi @ALL, here is the link to the LG webOS pre-test version:https://drive.google.com/file/d/1t9fRLavJAIzMPtsEb1h7Ox_C5IUjmnqQ/view?usp=sharing

Installed and it works a treat!! I have it running in a lamp using a philips hue bulb. I also have wled controller behind the TV controlling the strip

This is great work guys, much appreciated! Now add it to my other TV

**Comment by popy2k14** [comment]
> > There is an "transition" flag which could be given on color changes, so it will fade on it's own.
> 
> @popy2k14 I wanted to make you transition time available, but unfortunately the time can only be given in a multiple of one second, which I do not think makes sense. I have seen that for some devices, e.g. Yeelight one could define the time in ms in HA. For Hue I did not find that parameter...
> 
> **Transition**
> 
> > Number that represents the time (in seconds) the light should take to transition to the new state.
> 
> see https://www.home-assistant.io/integrations/light
> 
> @popy2k14 The error _"Selected Hyperion instance is not running"_ has been fixed too. It had nothing to do with cookies, but was caused by an inconsistent literal change.

Thanks for your ongoing improvements!

Transition time in HA is an decimal, so you can set it in HA like:

```
action: light.turn_on
data:
  transition: 0.9
target:
  entity_id: light.wohnzimmer
```

and in this example it would be 900ms.
So i think it would make sense to implement it that way.

**Comment by satgit62** [comment]
Hi @all,
I have noticed that via the “homeassistant” integration, some RGB lamps do not switch off in very dark or black scenes, but light up blue. Via the Philips Hue Entertainment integration, the signal detection timeouts for black have taken effect and the lamps have switched off. 
As a result, @Lord-Gray added the signal detection timeouts for black and installed an additional “turn off on black” switch. 
I have built a new LG webOS test version, so I can test this and see if it all works as intended. And it's great!

Here is the link:https://drive.google.com/file/d/1sQsFo6KARSjKNNKfOQCY4HVFYFoEEUw0/view?usp=sharing 

 
![Switch off on black](https://github.com/user-attachments/assets/0ad09201-5f5e-484d-a48d-90cf06388054)

**Comment by Lord-Grey** [comment]
> So i think it would make sense to implement it that way.

@popy2k14 I added a transition time for you for testing.
A transition time = 0 does not add "transition" and therefore does not override what is defined in HA.
Please play around with transition time > 0 and feedback, if I should define an lower or upper limit that is possible to configure.

**Comment by satgit62** [comment]
Hello, @all 
today @Lord-Grey has also added the checkbox “HA - Add transition time” under the “Switch off on black”, so that you can set it individually. This was a suggestion for improvement from @popy2k14 
Thanks again to @Lord-Grey for the quick implementation. Your feedback counts and helps @Lord-Gray to further develop the project. 
So test it out and let us know if everything works as desired.
For the LG webOS users, I have compiled the changes/commits of @Lord-Gray and built a test version. 
Download link: https://drive.google.com/file/d/1VNDHOREUw0oYW94PXasXL3NbXz8oTOee/view?usp=sharing   
----------------------------------------------------------------------------------------------------------------------------------------------
![Transition time](https://github.com/user-attachments/assets/01daba75-5752-4488-b431-6073d6aa349f)

**Comment by satgit62** [comment]
@all
Hello,

“Add Home Assistant Lights support” has been merged into Master with the following features:

* New HomeAssistant LED Device

* Fix typos

* Ping Qt for Windows to 6.7 until aqtinstaller is fixed

* Fix HA default port handling

* HA - Update default latch time and range

* Add HA Wizard and light selection

* Naming consistency

* Fix “Selected Hyperion instance is not running”

* CodeQL findings

* HA - allow overwriting brightness by HA yes or no

* HA - Support switch off on black

* HA - Add transition time


@all
Hi,

“Add Home Assistant Lights support” has been merged into Master with the following features:

I synchronized all Master changes in GitHub and built a new Hyperion.NG for LG webOS. Also, all effects are now available again. Furthermore, the ipk was reduced by 20 MB through the process of cleanup.

LG webOS Version Download Link: https://github.com/satgit62/hyperion.ng-webos-loader/releases/download/latest/org.webosbrew.hyperion.ng.loader_0.1.11_all.ipk

**Comment by kylewhirl** [comment]
How do we install this? I don’t see an update for Hyperion. Do we have to wait until it’s available?

**Comment by satgit62** [comment]
> How do we install this? I don’t see an update for Hyperion. Do we have to wait until it’s available?

Hi, where is your Hyperion installed? What are you running it on?

**Comment by kylewhirl** [comment]
I’m running it on a raspberry pi 3b with the official hyperbian image

**Comment by kylewhirl** [comment]
> > How do we install this? I don’t see an update for Hyperion. Do we have to wait until it’s available?
> 
> Hi, where is your Hyperion installed? On which platform are you running Hyperion?

I’m running it on a raspberry pi 3b with the official hyperbian image

**Comment by Lord-Grey** [comment]
@kylewhirl You can install a nightly build from our APT server to get the feature.

https://releases.hyperion-project.org/

**Comment by kylewhirl** [comment]
Thanks!On Aug 31, 2024, at 11:46 AM, LordGrey ***@***.***> wrote:﻿
@kylewhirl You can install a nightly build from our APT server to get the feature.
https://releases.hyperion-project.org/

**Comment by kylewhirl** [comment]
Just looked into the Home Assistant integration and looks like Hyperion can’t communicate with my HA server because my HA server is using https and Hyperion is trying to access it via http. Any way to fix this?

**Comment by Lord-Grey** [comment]
@kylewhirl Per HA documentation des API is a http one: https://developers.home-assistant.io/docs/api/rest/

Can you share Hyperion Debug logs?
Just curious what errors are reported.

**Comment by kylewhirl** [comment]
> @kylewhirl Per HA documentation des API is a http one: https://developers.home-assistant.io/docs/api/rest/
> 
> Can you share Hyperion Debug logs? Just curious what errors are reported.

The documentation does show http but that’s just the address of the server, so if the server uses https the api would also (confusing though cause they don’t mention this)

here are my logs:





8123]
2024-09-02T06:50:20.605Z [LEDDEVICE] (INFO) Resolved hostname (Home._home-assistant._tcp.local) to IP-address (192.168.xxx.xxx)
2024-09-02T06:50:20.617Z [LEDDEVICE] (DEBUG) (ProviderRestApi.cpp:274:executeOperation()) GET took 12ms, HTTP 0: [http://192.168.1.82:8123/api/states] []
2024-09-02T06:50:20.617Z [LEDDEVICE] (WARNING) homeassistant get properties failed with error: 'Connection closed'
2024-09-02T06:50:20.617Z [WEBSOCKET] (DEBUG) (JsonAPI.cpp:1275:handleLedDeviceGetProperties()) response: [{"properties":{}}]
2024-09-02T06:50:23.453Z [LEDDEVICE|Lamps] (DEBUG) (LedDevice.cpp:99:stop()) Stop device
2024-09-02T06:50:23.454Z [LEDDEVICE|Lamps] (DEBUG) (LedDevice.cpp:194:disable()) Disable device homeassistant'
2024-09-02T06:50:23.454Z [LEDDEVICE|Lamps] (INFO) Stopped LedDevice 'homeassistant'
2024-09-02T06:50:23.456Z [LEDDEVICE|Lamps] (INFO) Start LedDevice 'homeassistant'.
2024-09-02T06:50:23.457Z [LEDDEVICE|Lamps] (DEBUG) (LedDevice.cpp:215:init()) deviceConfig: [{"autoStart":true,"brightness":255,"colorOrder":"rgb","currentLedCount":1,"enableAttempts":6,"enableAttemptsInterval":15,"entityIds":[],"fullBrightnessAtStart":true,"hardwareLedCount":1,"host":"Home._home-assistant._tcp.local","hostList":"Home._home-assistant._tcp.local","latchTime":250,"overwriteBrightness":true,"port":8123,"restoreOriginalState":true,"switchOffOnBlack":true,"token":"redacted","transitionTime":0,"type":"homeassistant"}]

A checkbox to use SSL *should* work, I think

**Comment by Lord-Grey** [comment]
> A checkbox to use SSL should work, I think

@kylewhirl Yes, from a UI perspective it looks like a checkbox and the developer changes http to https.

…but https does not works without certificates.
Therefore, what certificates are you using, where do you store the public part of it and how have you configured HA for connectivity.

If you want, you could ping me the details via a PM on Discord or the Hyperion forum to not share details publicly.

**Comment by Adesfire** [comment]
Hi there! 
How do I get this version compiled for Windows? I can only find the 2.0.16 so far...
Thank you in advance!

**Comment by Lord-Grey** [comment]
@adesfire See https://github.com/hyperion-project/hyperion.ng/blob/master/doc/development/CompileHowto.md#windows

Or 

Check if there was a PR lately build and download the attached Windows build.
 Currently, [this one](https://github.com/hyperion-project/hyperion.ng/actions/runs/12574181472) has a ready made build still attached.

**Comment by cpaulik** [comment]
Does anybody have a link to the latest web os version that includes transition time?

I tried installing

- https://github.com/Paulchen-Panther/hyperion.ng-webos-loader/releases/tag/latest
- https://github.com/satgit62/hyperion.ng-webos-loader/releases/tag/latest

but got this error

<img width="643" alt="Image" src="https://github.com/user-attachments/assets/256ce836-0b7d-4d80-acb4-329e3af8b69e" />

**Comment by satgit62** [comment]
> Does anybody have a link to the latest web os version that includes transition time?
> 
> I tried installing
> 
>     * https://github.com/Paulchen-Panther/hyperion.ng-webos-loader/releases/tag/latest
> 
>     * https://github.com/satgit62/hyperion.ng-webos-loader/releases/tag/latest
> 
> 
> but got this error
> <img alt="Image" width="643" src="https://private-user-images.githubusercontent.com/380927/437234392-256ce836-0b7d-4d80-acb4-329e3af8b69e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDU1NjQ5NTcsIm5iZiI6MTc0NTU2NDY1NywicGF0aCI6Ii8zODA5MjcvNDM3MjM0MzkyLTI1NmNlODM2LTBiN2QtNGQ4MC1hY2I0LTMyOWUzYWY4YjY5ZS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNDI1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDQyNVQwNzA0MTdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0xN2RiOWEyOTE3MTRjZGVjMTBkNjNkZDQ5OTEwNTBkYmYwYmEyNTEzOGU0MTZmYWJhNDVhY2U1ZjMzZmZkNzJmJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.5D4J7otHIlxZ-ys3SKkryktsLmZPDpHoJS8_vvKWhLU">

Hello, have you tried to install using the webOS Dev Manager?
In the past, there have been problems with the certificate check in the homebrew channel or setting the date and time.

Installing an IPK may fail with an error like this:
```json
{"id":"com.ares.defaultName","statusValue":24,"details":{"receivedSize":"0","packageId":"","modifiedTime":"","unpackFileSize":"0","client":"com.webos.lunasendpub-1234","totalSize":"0","verified":false,"progress":100,"errorCode":-5,"installBasePath":"/media/developer","reason":"ipk verified failed","downgrade":true,"paused":false,"permissionLevel":"","simpleStatus":"install","state":"install failed","pkgType":"","update":false,"ipkFile":"/path/to/app.ipk"}}```

Make sure your TV's date is set (approximately) correctly. Specifically, ensure that the TV's date is newer than the dates of the files in the IPK. 

Note that if you are blocking the SDP server, the TV won't set the date/time on boot. In such cases, the date usually defaults to the beginning of the year in which the TV model was introduced.

Here is the latest version of Hyperion.NG: https://github.com/satgit62/hyperion.ng-webos-loader/releases/download/latest/org.webosbrew.hyperion.ng.loader_0.1.12_all.ipk

**Comment by satgit62** [comment]
> Does anybody have a link to the latest web os version that includes transition time?
> 
> I tried installing
> 
>     * https://github.com/Paulchen-Panther/hyperion.ng-webos-loader/releases/tag/latest
> 
>     * https://github.com/satgit62/hyperion.ng-webos-loader/releases/tag/latest
> 
> 
> but got this error
> <img alt="Image" width="643" src="https://private-user-images.githubusercontent.com/380927/437234392-256ce836-0b7d-4d80-acb4-329e3af8b69e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDU1NjQ5NTcsIm5iZiI6MTc0NTU2NDY1NywicGF0aCI6Ii8zODA5MjcvNDM3MjM0MzkyLTI1NmNlODM2LTBiN2QtNGQ4MC1hY2I0LTMyOWUzYWY4YjY5ZS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNDI1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDQyNVQwNzA0MTdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0xN2RiOWEyOTE3MTRjZGVjMTBkNjNkZDQ5OTEwNTBkYmYwYmEyNTEzOGU0MTZmYWJhNDVhY2U1ZjMzZmZkNzJmJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.5D4J7otHIlxZ-ys3SKkryktsLmZPDpHoJS8_vvKWhLU">

Hello, have you tried to install using the webOS Dev Manager?
In the past, there have been problems with the certificate check in the homebrew channel or setting the date and time.

Installing an IPK may fail with an error like this:
```json
{"id":"com.ares.defaultName","statusValue":24,"details":{"receivedSize":"0","packageId":"","modifiedTime":"","unpackFileSize":"0","client":"com.webos.lunasendpub-1234","totalSize":"0","verified":false,"progress":100,"errorCode":-5,"installBasePath":"/media/developer","reason":"ipk verified failed","downgrade":true,"paused":false,"permissionLevel":"","simpleStatus":"install","state":"install failed","pkgType":"","update":false,"ipkFile":"/path/to/app.ipk"}}```

Make sure your TV's date is set (approximately) correctly. Specifically, ensure that the TV's date is newer than the dates of the files in the IPK. 

Note that if you are blocking the SDP server, the TV won't set the date/time on boot. In such cases, the date usually defaults to the beginning of the year in which the TV model was introduced.

Here is the latest version of Hyperion.NG: https://github.com/satgit62/hyperion.ng-webos-loader/releases/download/latest/org.webosbrew.hyperion.ng.loader_0.1.12_all.ipk

**Comment by cpaulik** [comment]
thanks. 

> Hello, have you tried to install using the webOS Dev Manager?

Yes that is what I tried.

I tried this now:

```
root@LGwebOSTV:~# luna-send-pub -i 'luna://org.webosbrew.hbchannel.service/install' '{"id":"com.ares.defaultName","ipkUrl":"https://github.com/satgit62/hyperion.ng-webos-loader/releases/download/latest/org.webosbrew.hyperion.ng.loader_0.1
.12_all.ipk","subscribe":false, "ipkHash":"6cd3bdd848a7e782ec3ee2bf4590cb291421f97f3ddc75ba3899d90fef613801"}'
{"statusText":"Downloading…","returnValue":true}
{"statusText":"Downloading…","progress":13.013676617977362,"returnValue":true}
{"statusText":"Downloading…","progress":26.152347961646615,"returnValue":true}
{"statusText":"Downloading…","progress":46.11419221393337,"returnValue":true}
{"statusText":"Downloading…","progress":65.34239205012557,"returnValue":true}
{"statusText":"Downloading…","progress":83.60071676663013,"returnValue":true}
{"statusText":"Downloading…","progress":100,"returnValue":true}
{"statusText":"Verifying…","returnValue":true}
{"statusText":"Installing…","returnValue":true}
{"returnValue":false,"errorText":"-5: ipk verified failed"}
cc
^C
root@LGwebOSTV:~# ^C
root@LGwebOSTV:~# date
Fri Apr 25 10:54:00 CEST 2025
```

**Comment by satgit62** [comment]
@cpaulik 

Hi,
What happens if they first with
 ```curl -k -L -o /tmp/app.ipk 'https://github.com/satgit62/hyperion.ng-webos-loader/releases/download/latest/org.webosbrew.hyperion.ng.loader_0.1.12_all.ipk'``` and then with ```luna-send-pub -w 15000 -i 'luna://com.webos.appInstallService/dev/install' '{"id": "com.ares.defaultName", "ipkUrl":"/tmp/app.ipk", "subscribe":true}'```, install.

**Comment by cpaulik** [comment]
Same thing

```
luna-send-pub -w 15000 -i 'luna://com.webos.appInstallService/dev/install' '{"id": "com.ares.defaultName", "ipkUrl":"/tmp/app.ipk", "subscribe":true}'
{"subscribed":true,"returnValue":true}
{"id":"com.ares.defaultName","statusValue":268,"details":{"receivedSize":"0","packageId":"","modifiedTime":"","unpackFileSize":"0","client":"com.webos.lunasendpub-6076","totalSize":"0","verified":false,"progress":100,"installBasePath":"/media/developer","downgrade":true,"paused":false,"permissionLevel":"","simpleStatus":"install","state":"ipk verifying","pkgType":"","update":false,"ipkFile":"/tmp/app.ipk"}}
{"id":"com.ares.defaultName","statusValue":269,"details":{"receivedSize":"0","packageId":"","modifiedTime":"","unpackFileSize":"0","client":"com.webos.lunasendpub-6076","totalSize":"0","verified":false,"progress":100,"installBasePath":"/media/developer","downgrade":true,"paused":false,"permissionLevel":"","simpleStatus":"install","state":"ipk verifying","pkgType":"","update":false,"ipkFile":"/tmp/app.ipk"}}
{"id":"com.ares.defaultName","statusValue":24,"details":{"receivedSize":"0","packageId":"","modifiedTime":"","unpackFileSize":"0","client":"com.webos.lunasendpub-6076","totalSize":"0","verified":false,"progress":100,"errorCode":-5,"installBasePath":"/media/developer","reason":"ipk verified failed","downgrade":true,"paused":false,"permissionLevel":"","simpleStatus":"install","state":"install failed","pkgType":"","update":false,"ipkFile":"/tmp/app.ipk"}}
```

syslog
```
[11:52:10.809][ERR] pkgVerifier ERRMSG PkgVerifierLunaService::extractIPK FILE:tar: control: implausibly old time stamp 1970-01-01 01:00:00 ---3d 0x7d8bd0 {}
[11:52:10.810][ERR] pkgVerifier ERRMSG PkgVerifierLunaService Control file Extraction Failure {}
[11:52:10.812][ERR] AppInstallD TASK_ERROR  {"app_id":"com.ares.defaultName","error_code":-5,"error_text":"ipk verified failed"}
[11:52:10.818][WARNING] AppInstallD IPK_VERIFY_ERR Ipk verify failed {"app_id":"com.ares.defaultName","error_text":"ipk verified failed"}
```

**Comment by satgit62** [comment]
> Same thing
> 
> ```
> luna-send-pub -w 15000 -i 'luna://com.webos.appInstallService/dev/install' '{"id": "com.ares.defaultName", "ipkUrl":"/tmp/app.ipk", "subscribe":true}'
> {"subscribed":true,"returnValue":true}
> {"id":"com.ares.defaultName","statusValue":268,"details":{"receivedSize":"0","packageId":"","modifiedTime":"","unpackFileSize":"0","client":"com.webos.lunasendpub-6076","totalSize":"0","verified":false,"progress":100,"installBasePath":"/media/developer","downgrade":true,"paused":false,"permissionLevel":"","simpleStatus":"install","state":"ipk verifying","pkgType":"","update":false,"ipkFile":"/tmp/app.ipk"}}
> {"id":"com.ares.defaultName","statusValue":269,"details":{"receivedSize":"0","packageId":"","modifiedTime":"","unpackFileSize":"0","client":"com.webos.lunasendpub-6076","totalSize":"0","verified":false,"progress":100,"installBasePath":"/media/developer","downgrade":true,"paused":false,"permissionLevel":"","simpleStatus":"install","state":"ipk verifying","pkgType":"","update":false,"ipkFile":"/tmp/app.ipk"}}
> {"id":"com.ares.defaultName","statusValue":24,"details":{"receivedSize":"0","packageId":"","modifiedTime":"","unpackFileSize":"0","client":"com.webos.lunasendpub-6076","totalSize":"0","verified":false,"progress":100,"errorCode":-5,"installBasePath":"/media/developer","reason":"ipk verified failed","downgrade":true,"paused":false,"permissionLevel":"","simpleStatus":"install","state":"install failed","pkgType":"","update":false,"ipkFile":"/tmp/app.ipk"}}
> ```
> 
> syslog
> 
> ```
> [11:52:10.809][ERR] pkgVerifier ERRMSG PkgVerifierLunaService::extractIPK FILE:tar: control: implausibly old time stamp 1970-01-01 01:00:00 ---3d 0x7d8bd0 {}
> [11:52:10.810][ERR] pkgVerifier ERRMSG PkgVerifierLunaService Control file Extraction Failure {}
> [11:52:10.812][ERR] AppInstallD TASK_ERROR  {"app_id":"com.ares.defaultName","error_code":-5,"error_text":"ipk verified failed"}
> [11:52:10.818][WARNING] AppInstallD IPK_VERIFY_ERR Ipk verify failed {"app_id":"com.ares.defaultName","error_text":"ipk verified failed"}
> ```

It's strange that it installs on my device but not on hers.
I unzipped the IPK and sure enough, the timestamp is 1970-01-01. I'll have to check the repo again to see why that is, since Hyperion.NG was reprogrammed from scratch.
Is the system time on the TV correct?

**Comment by cpaulik** [comment]
Yeah, really strange. 

> Is the system time on the TV correct?

Yes, It was set automatically but I also tried setting it manually.

**Comment by cpaulik** [comment]
For now I seem to have fixed my problem by re-packaging the ipk file

```
tar -xf org.webosbrew.hyperion.ng.loader_0.1.12_all.ipk
tar -xf control.tar.gz
tar -xf data.tar.gz
find . -exec touch {} +
rm control.tar.gz
rm data.tar.gz
gtar --format=gnu --owner=0 --group=0 --numeric-owner -czf control.tar.gz control
gtar --format=gnu --owner=0 --group=0 --numeric-owner -czf data.tar.gz usr
ar cru fixed_app.ipk debian-binary control.tar.gz data.tar.gz
```

Installing `fixed_app.ipk` works.

Was the HA transition time removed again. I don't see it as an option?

**Comment by satgit62** [comment]
@cpaulik 
Hello,
here is the latest build of Hyperion.NG for download.
I hope the installation now works for you too, without any errors.
https://github.com/satgit62/hyperion.ng-webos-loader/actions/runs/14714298229/artifacts/3024036135
